tests/libwnbd_tests/request_log.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2022 Cloudbase Solutions
+ *
+ * Licensed under LGPL-2.1 (see LICENSE)
+ */
+
+#include "pch.h"
+
+#include "request_log.h"
+
+void RequestLog::PushBack(const RequestLogEntry& Entry)
+{
+	std::unique_lock<std::mutex> Lock(ListLock);
+
+	auto RequestHandle = Entry.WnbdRequest.RequestHandle;
+
+	// Validate the request handle.
+	auto Result = RequestHandles.insert(RequestHandle);
+	ASSERT_TRUE(Result.second)
+		<< "duplicate request handle received : "
+		<< RequestHandle;
+
+	Entries.push_back(Entry);
+}
+
+void RequestLog::AddEntry(
+	WNBD_IO_REQUEST& WnbdRequest,
+	void* DataBuffer,
+	size_t DataBufferSize)
+{
+	void* BufferCopy = nullptr;
+	DWORD BufferCopySize = 0;
+
+	if (DataBuffer && DataBufferSize > 0) {
+		BufferCopySize = DataBufferSize;
+		BufferCopy = malloc(BufferCopySize);
+		ASSERT_TRUE(BufferCopy)
+			<< "couldn't allocate buffer, size: " << BufferCopySize;
+		memcpy(BufferCopy, DataBuffer, BufferCopySize);
+	}
+
+	RequestLogEntry Entry(WnbdRequest, BufferCopy, BufferCopySize);
+	PushBack(Entry);
```

Comment:
should the entry be pushed even when BufferCopy is equal to nullptr?

Reply:
Sure, the data buffer is optional. for example, it might be a flush or a read.